File: C++/reverse-words-in-a-string-ii.cpp
class Solution {
private:
    void reverse_word(std::string &s, int start, int end) {
        end -= 1;
        for(; start<end; ++start, --end) {
            std::swap(s[start], s[end]);
        }
    }
public:
    void reverseWords(std::string &s) {
        int start = 0;
        int end = 0;
        for(; end<s.size(); ++end) {
            if(s[end]==' ') {
                reverse_word(s, start, end);
                start = end+1;
            }
        }
        reverse_word(s, start, end);
        reverse_word(s, 0, s.size());
    }
};

File: Python/reverse-words-in-a-string-ii.py
class Solution:
    # @param s, a list of 1 length strings, e.g., s = ['h','e','l','l','o']
    # @return nothing
    def reverseWords(self, s):
        def reverse_substr(s, start, end):
            end -= 1
            while start<end:
                s[start], s[end] = s[end], s[start]
                start += 1
                end -= 1
        
        n = len(s)
        reverse_substr(s, 0, n)
        
        start = 0
        
        for i in xrange(n):
            if s[i] == ' ':
                reverse_substr(s, start, i)
                start = i+1
        
        reverse_substr(s, start, n)